### PR TITLE
refactor: use `class` as prop to prevent class duplication with `cn` function

### LINKED ISSUE #35: [Bug]: Class name duplicated
If you inspecting some element, you would notice a duplicated class name mainly due to using `2.` approach mentioned below.

For components in `registry`, you would see 2 different approach for passing the class name into `cn` function. 

1. using `class` props.
```
const props = defineProps<... & { class?: string }>()
...
cn(...., props.class)
```

2. using inherited attributes
```
cn(..., $attrs.class)
```

The reason I'm hesitating is because I don't like to declare `class` props in every component, using `$attrs.class` seems much more intuitive. wdyt??

### LINKED ISSUE #310: [Bug]: Uncontrolled collapsible not working
### Environment

```bash
Developement/Production OS: Windows 10 Pro 10.0.19045
Node version: 18.14.0
Package manager: pnpm@8.12.1
Radix Vue version: 1.3.2
Shadcn Vue version: 0.8.7
Vue version: 3.4.0
Nuxt version: 3.0.0
Nuxt mode: universal
Nuxt target: server
CSS framework: tailwindcss@3.4.0
Client OS: Windows 10 Pro 10.0.19045
Browser: Chrome 120.0.6099.130
```

### Link to minimal reproduction

https://stackblitz.com/edit/w2uuxx-s8mgnp?file=src%2FApp.vue,package.json

### Steps to reproduce

1. Start a new Vue 3, Typescript project with Vite and create-vue `pnpm create vue@latest`.
2. Install tailwindcss and its peer dependencies.
3. After that Run the shadcn-vue init command to setup shadcn-vue in project `pnpm dlx shadcn-vue@latest init`.
4. Install Collapsible and Button component from shadcn-vue.
5. Start the local dev server `pnpm run dev`.
6. Create two Collapsible one with `ref` (Controlled) and Second (Uncontrolled).

### Describe the bug

I was checking the docs on mobile the other day and found out that on mobile "On this Page" TOC was not working. So, i decided to check what is the issue?

[on-this-page-shadcn-vue.webm](https://github.com/radix-vue/shadcn-vue/assets/62164346/cc14ca0f-0610-4a24-a4af-0d6c3856f03b)

In the docs we are using TableOfContent.vue component and it will show TOC for the page and there we have this [On This Page Collapsible](https://github.com/radix-vue/shadcn-vue/blob/dev/apps/www/.vitepress/theme/components/TableOfContent.vue#L66) which will toggle the TOC section.

The code had no issue because we can use Collapsible both Controlled and Uncontrolled way. Here, In the docs we are using as Uncontrolled and it should work. Then i checked in our [Collapsible Component Docs](https://www.shadcn-vue.com/docs/components/collapsible.html) and there it was working but not for the "On this page" TOC.

Controlled collapsible is working but Uncontrolled collapsible not working. Also i am not sure whether this is bug from radix-vue side from our collapsible component side.

[uncontrolled-collapsible-not-working.webm](https://github.com/radix-vue/shadcn-vue/assets/62164346/4684df83-c527-405a-bcdd-98d9eae557be)

### Expected behavior

Collapsible should work if its controlled or uncontrolled.

### Conext & Screenshots (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by sadeghbarati**
close #35 
close #310 

### This might take a bit longer cause some of the classes/styles changed in main shadcn-ui  so I have to bring them in here too

---

Previously we used destructure `useAttrs` like 👇 , to fix class `duplication`

```ts
defineOption({
  inheritAttrs: false
})

const { class: className, ...rest } = useAttrs()
```

But destructure props or attrs will cause lose reactivity

We could make `attrs` reactive with `computed` or `toRefs`, 
But making **Always read-only (inside the component)** thing, reactive/mutable doesn't make sense 

image

---

Now we are using class in prop and delegated/rest

<details><summary>delegated/rest spread props</summary>
<p>

```vue
<script setup lang="ts">
import { type HTMLAttributes, computed } from 'vue'
import { AccordionContent, type AccordionContentProps } from 'radix-vue'
import { cn } from '@/lib/utils'

const props = defineProps<AccordionContentProps & { class?: HTMLAttributes['class'] }>()

const delegatedProps = computed(() => {
  const { class: _, ...delegated } = props

  return delegated
})
</script>

<template>
  <AccordionContent
    v-bind="delegatedProps"
    :class="
      cn(
        'overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down',
        props.class,
      )
    "
  >
    <div class="pb-4 pt-0">
      <slot />
    </div>
  </AccordionContent>
</template>
```

</p>
</details> 

---

- After this rfcs https://github.com/vuejs/rfcs/discussions/502 comes out of **experimental** we will have this code

  <details><summary>Details</summary>
  <p>
  
  ```vue
  <script setup lang="ts">
  import type { HTMLAttributes } from 'vue'
  import { AccordionContent, type AccordionContentProps } from 'radix-vue'
  import { cn } from '@/lib/utils'
  
  const { class: className, ...delegated  } = defineProps<AccordionContentProps & { class?: HTMLAttributes['class'] }>()
  </script>
  
  <template>
    <AccordionContent
      v-bind="delegated "
      :class="
        cn(
          'overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down',
          className,
        )
      "
    >
      <div class="pb-4 pt-0">
        <slot />
      </div>
    </AccordionContent>
  </template>
  ```
  
  </p>
  </details> 

- And after complex types support for defineProps, we will have this, hopefully xd

  <details><summary>Details</summary>
  <p>
  
  ```vue
  <script setup lang="ts">
  import type { HTMLAttributes } from 'vue'
  import { AccordionContent, type AccordionContentProps } from 'radix-vue'
  import { cn } from '@/lib/utils'
  
  const { class: className, ...delegated  } = defineProps<AccordionContentProps & HTMLAttributes>()
  </script>
  
  <template>
    <AccordionContent
      v-bind="delegated "
      :class="
        cn(
          'overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down',
          className,
        )
      "
    >
      <div class="pb-4 pt-0">
        <slot />
      </div>
    </AccordionContent>
  </template>
  ```
  
  </p>
  </details> 

- Complex types + generics (not real code 😂) for example take look at this react article 

https://www.totaltypescript.com/pass-component-as-prop-react

  <details><summary>Details</summary>
  <p>
  
  ```vue
  <script setup lang="ts" generic="T extends AsTag">
  import type { IntrinsicElementAttributes } from 'vue'
  import { AccordionContent, type AccordionContentProps } from 'radix-vue'
  import { cn } from '@/lib/utils'
  
  const { class: className, ...delegated  } = defineProps<AccordionContentProps & IntrinsicElementAttributes[T]>()
  </script>
  
  <template>
    <AccordionContent
      v-bind="delegated "
      :class="
        cn(
          'overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down',
          className,
        )
      "
    >
      <div class="pb-4 pt-0">
        <slot />
      </div>
    </AccordionContent>
  </template>
  ```
  
  </p>
  </details>

### CODE REVIEW

**Comment by sadeghbarati** [comment]
This PR **Depends on** this change 👇  in `radix-vue` to be released

https://github.com/radix-vue/radix-vue/pull/628


---

### Please **`remove the commit body`** if you want to Merge it into dev 🤣 

<details><summary>Details</summary>
<p>

image

</p>
</details>

**Comment by sadeghbarati** [comment]
There is an error with `codeeditor` and `CodeSandbox.vue` or `Stackblitz.vue` playgrounds which IDK what is the problem

Can you please check it? @zernonia 

Thanks

**Comment by sadeghbarati** [comment]
Manually patched radix-vue to `v1.4.0` in my local Stackblitz and Codesandbox playgrounds is working fine


There is some bug with the new `radix-vue` Popover in `1.4.0` in Menubar

https://github.com/radix-vue/shadcn-vue/assets/17789047/c9c6c8b9-5de1-4c89-8fe0-bd323b6cea49

**Comment by sadeghbarati** [comment]
radix-vue `v1.4.0` Popover issue

https://github.com/radix-vue/shadcn-vue/assets/17789047/2ca2278c-d0b9-4569-b9d7-900e2e7f4e61

File: apps/www/src/lib/registry/default/ui/accordion/AccordionContent.vue [hunk 1]
```diff
@@ -1,21 +1,23 @@
 <script setup lang="ts">
+import { type HTMLAttributes, computed } from 'vue'
 import { AccordionContent, type AccordionContentProps } from 'radix-vue'
 import { cn } from '@/lib/utils'
 
-const props = defineProps<AccordionContentProps & { class?: string }>()
+const props = defineProps<AccordionContentProps & { class?: HTMLAttributes['class'] }>()
```
Comment (zernonia) [inline apps/www/src/lib/registry/default/ui/accordion/AccordionContent.vue:6]: I worry this `HTMLAttributes` would cause some issue for `detypes` 😂

**Review by zernonia — APPROVED** [review]

**Comment by zernonia** [comment]
@sadeghbarati  do upgrade the `radix-vue` deps to `v1.4.0`

**Comment by hendrasan** [comment]
I'm not sure whether this is the right place or not, but i found some bugs that I traced back to this commit.

I'm using vue 3.3.8, haven't upgraded to vue 3.4 yet.

So in SelectItem.vue, the problem is in this code

```
const props = defineProps<SelectItemProps & { class?: HTMLAttributes['class'] }>()

const delegatedProps = computed(() => {
  const { class: _, ...delegated } = props

  return delegated
})

const forwardedProps = useForwardProps(delegatedProps)
console.log('f', forwardedProps)
```
If I'm using radix-vue version 1.2.3, this breaks as the type of forwardedProps become CustomRefImpl when I console.log it, and the value of it becomes an object like this:

```
value: {
    "value": "pineapple",
    "disabled": false,
    "asChild": false
}
```
To get around this, I tested two solutions:

1. Upgrade radix-vue to 1.4.0, then the type of forwardedProps become the expected ComputedRefImpl and the value is the expected string value, which is "pineapple" in the above example. But in radix-vue changelog, it said that v1.3.0 required vue 3.4 (not sure if it's just for the treeshaking feature, I tested radix-vue 1.4.0 in this vue 3.3.8 app and it seemed to still work fine, haven't tested the whole app)
2. Stay in radix-vue 1.2.3, don't use the forwardedProps, but directly use the props to pass to v-bind in SelectItem component, not sure if this is a good workaround though.

Also, whenever i add a component using `npx shadcn-vue@latest add select`, it prompts me to upgrade the shadcn-vue version, but IIRC the radix-vue package is installed only during `npx shadcn-vue@latest init`. So if there are changes that require updates to radix-vue like this one, do I have to manually install the updated radix-vue or can it be automated in some way?

**Pull request merged**